### DISCUSSION TITLE
FYI: a developer begun porting ESPHome to nrf52/Zephyr for nRF52840 and started development on a Zigbee component for ESPHome based on ZBOSS

### DISCUSSION THREAD

**Discussion by Hedda** [ffenix113/zigbee_home#23] [General]
Off-topic but FYI; @tomaszduda23 has started working on an initial (basic) port of [ESPHome](https://esphome.io/) for nRF Connect / nRF5 SDK (for nRF52840) as well as Zigbee support (for nRF) on that platform. See these pull requests for ESPHome which contain summeries and detailed descriptions:

- https://github.com/esphome/esphome/pull/7049
- https://github.com/esphome/esphome/pull/11535
  - https://github.com/esphome/esphome/pull/12187
  - https://github.com/esphome/esphome/pull/11535
  - https://github.com/esphome/esphome/pull/11533
  - https://github.com/esphome/esphome/pull/13083
  - https://github.com/esphome/esphome/pull/13114
  - https://github.com/esphome/esphome/pull/12236

**Update:** ESPHome 2026.1.0 has now been released with a Zigbee End Device component based on nRF SDK supporting nRF52 platform:

* https://esphome.io/components/zigbee/

"_**The zigbee component allows exposing supported ESPHome components over a Zigbee network to Home Assistant via Zigbee2MQTT or ZHA. Due to the limitations of the Zigbee protocol, only basic properties are exposed. Additional properties must be configured manually in Home Assistant. Each ESPHome entity consumes one Zigbee endpoint. Because of a limitation in Zigbee2MQTT, at least two endpoints are required. The maximum number of supported endpoints is eight.**_"

[tomaszduda23](https://github.com/tomaszduda23) is currently activly developing most of it in his "nrf52_zigbee" and "nrf52" branches in his forked repo of ESPHome:

- https://github.com/tomaszduda23/esphome/tree/nrf52_zigbee

  - https://github.com/tomaszduda23/esphome/tree/nrf52

    - https://github.com/esphome/esphome/pull/7340

Related as he made short a post informing about this in ESPHome's feature request discussion about adding Zigbee support to ESPHome: 

https://github.com/esphome/feature-requests/issues/1397#issuecomment-1922574971

  "**_I've started porting esphome to nRF52840 a few weeks ago https://github.com/esphome/esphome/pull/6075. It can be used with nrf-sdk which has some zigbee examples https://github.com/nrfconnect/sdk-nrf/tree/main/samples/zigbee. I've not looked in zigbee details. Maybe someone would be interested though._**"

  "_**It adds zigbee switch and binary_sensor. Tested with zigbee2mqtt 1.40.0. Up to 8 components is supported due to endpoint limit on nrf52840. Endpoints could be reused but it requires extra implementation. Entities names are auto discovered by home assistant.**_"

That development effort might also be an interesting sub-project to follow for those who (like myself) have been asking for native Zigbee  support in ESPHome (and perhaps posted in the feature request tracker) since Nordic Semiconductor nRF52840 does not support WiFi but does support both Zigbee and Thread protocols, (similar to the ESP32-H2, while Wspressif also had ESP32-C5 and ESP32-C6 with WiFi too).

**Comment by Hedda** [comment]
@ffenix113 still off-topic but another indirectly related FYI; @tomaszduda23 has now created a Zigbee proof-of-concept for ESPHome on nRF52840 with YAML config here -> https://github.com/tomaszduda23/esphome/pull/2

**Reply by Hedda** [reply]
@ ffenix113 FYI, the same developer have now submitted a PR for initialZigbee support to ESPHome

- https://github.com/esphome/esphome/pull/7340

That in turn require his PR to add nRF52 support to Zephyr for ESPHome:

- https://github.com/esphome/esphome/pull/7049

**Comment by Hedda** [comment]
FYI, [tomaszduda23](https://github.com/tomaszduda23) keep making great progress towards the path to add both nrf52 platform core (based on Zephyr) and Zigbee for nRF to the ESPHome firmware (developed for the nRF52840 MCU SoC chip):

- https://github.com/esphome/esphome/pull/7340

- https://github.com/esphome/esphome/pull/7049

That custom nrf52 platform be an external component:

- https://github.com/esphome/esphome/pull/7616

PS: His previous Zigbee proof-of-concept example for ESPHome on nRF52840 with YAML config for reference here:

- https://github.com/tomaszduda23/esphome/pull/2

**Comment by Hedda** [comment]
@ffenix113 FYI, initial pull requests for nrf52 core (based on zephyr) from [tomaszduda23](https://github.com/tomaszduda23) has now been merged into ESPHome dev branch:

* https://github.com/esphome/esphome/pull/7049

plus many more new related nrf52 patches are in development now that the initial support has been added and will soon be in the mainline:

* https://github.com/esphome/esphome/pulls?q=is%3Apr+nrf52+

Again more revevant is that an initial pull request for Zigbee support on nrf52 for ESPHome is now also getting close to being merged soon:

* https://github.com/esphome/esphome/pull/7340

PS: Also noted that [luar123](https://github.com/luar123) has begun working on an external low-level Zigbee component support for ESPHome on Espressif ESP32 as well:

* https://github.com/luar123/zigbee_esphome

**Comment by Hedda** [comment]
@ffenix113 FYI, ESPHome 2025.8.0 release from the 20th of August 2025 has now added official support for nRF52 as a new platform:

* https://esphome.io/changelog/2025.8.0.html

This means that there is now native support in ESPHome for nRF52840 boards (based on Zephyr), and while it is still early it is a great start:

  * https://esphome.io/components/nrf52

Note though that Zigbee is still not support in mainline but can be tested via his fork or by loading it as external component for ESPHome:

* https://github.com/esphome/esphome/pull/7340

  * https://github.com/tomaszduda23/esphome/tree/nrf52_zigbee

It currenly requires an Adafruit bootloader (example `adafruit_nrf52_sd140_v6` or `adafruit_itsybitsy_nrf52840` depending on board used).

# _NRF52 Platform Support_

_ESPHome 2025.8.0 introduces basic support for Nordic nRF52 series microcontrollers through the new [NRF52 Platform](https://esphome.io/components/nrf52) platform. This implementation is built on the Zephyr RTOS, providing a robust foundation for Nordic semiconductor devices._

_Key Features:_

* _**Complete platform integration** with ESPHome’s component ecosystem_
* _**ADC support** for analog sensor reading with configurable resolution_
* _**GPIO functionality** with interrupt support and pin configuration_
* _**Zephyr debug component** for advanced debugging and development_

_The nRF52 platform opens up new possibilities for low-power, Bluetooth-enabled devices while maintaining the familiar ESPHome configuration syntax and component compatibility._

------------------------------

This is a lead up to that this pull requestsfor initial Zigbee device support has been submitted to the upstream ESPHome firmware repo:

- https://github.com/esphome/esphome/pull/7340

As well as many other related pull requests that adds an underlying framework or more extensive debugging support for nRF52 and Zephyr:

 - https://github.com/esphome/esphome/pull/8150
 - https://github.com/esphome/esphome/pull/9319
 - https://github.com/esphome/esphome/pull/8319
 - https://github.com/esphome/esphome/pull/8315
 - https://github.com/esphome/esphome/pull/8992
 - https://github.com/esphome/esphome/pull/8352
 - https://github.com/esphome/esphome/pull/7715
 - https://github.com/esphome/esphome/pull/8139
 - https://github.com/esphome/esphome/pull/8321
 - https://github.com/esphome/esphome/pull/8412
 - https://github.com/esphome/esphome/pull/8150
 - https://github.com/esphome/esphome/pull/9321
 - https://github.com/esphome/esphome/pull/9859
 - https://github.com/esphome/esphome/pull/9858
 - https://github.com/esphome/esphome/pull/9864
 - https://github.com/esphome/esphome/pull/10698
 - https://github.com/esphome/esphome/pull/9846
 - https://github.com/esphome/esphome/pull/9861
 - https://github.com/esphome/esphome/pull/10591
 - https://github.com/esphome/esphome/pull/10694
 - https://github.com/esphome/esphome/pull/10695
 - https://github.com/esphome/esphome/pull/11672
 - https://github.com/esphome/esphome/pull/11839
 - https://github.com/esphome/esphome/pull/11841
 - https://github.com/esphome/esphome/pull/11846
 - https://github.com/esphome/esphome/pull/11847
 - https://github.com/esphome/esphome/pull/11854
 - https://github.com/esphome/esphome/pull/11867
 - https://github.com/esphome/esphome/pull/11931
 - https://github.com/esphome/esphome/pull/11932 
 - https://github.com/esphome/esphome/pull/11933
 - https://github.com/esphome/esphome/pull/12053
 - https://github.com/esphome/esphome/pull/12344
 - https://github.com/esphome/esphome/pull/12489
 - https://github.com/esphome/esphome/pull/12582
 - https://github.com/esphome/esphome/pull/13022
 - https://github.com/esphome/esphome/pull/13122
 - https://github.com/esphome/esphome/pull/13249
 - https://github.com/esphome/esphome/pull/13317
 - https://github.com/esphome/esphome/pull/13333

------------------------------

PS: Again. there is also another developer working on Zigbee support for ESHome on ESP32 platform as well so will see if both gets merged, though that other one by luar123 uses a much more low-level approch than what tomaszduda23 is working on for the nRF52 platform:

- https://github.com/luar123/zigbee_esphome

**Comment by Hedda** [comment]
Update: still experimental but ESPHome 2025.9.0 and onwards contain many added capabilities and enchancements to their nRF52 platform:

* https://esphome.io/components/nrf52/

  * https://esphome.io/changelog/2025.9.0/

  * https://esphome.io/changelog/2025.10.0/

  * https://esphome.io/changelog/2025.11.0/

FYI, tomaszduda23 also continues to refine his pull request to add Zigbee support for the nRF52 platform in ESPHome which show potential:

* https://github.com/esphome/esphome/pull/7340

Note; the nRF52 ZBOSS Zigbee PR currently only supports Zigbee binary sensor, switch, time, sensor, and number as ESPHome device types. And tomaszduda23 mentioned there that he is currently not planing to implement any others clusters but said that PRs are welcome though:

  * https://github.com/tomaszduda23/esphome/tree/nrf52_zigbee

PS: Perhaps also relevant is that the next ESPHome version should make it easier to use ESP32 Arduino as an ESP-IDF component (as in ESPHome 2025.10.0 release and later will have a breaking change migrating the use of the ESP32 Arduino framework to use Arduino as an ESP-IDF component, so while off-topic here wondering if that mainline ESPHome change to use ESP32 Arduino as an ESP-IDF component would also change how to the mentioned [external Zigbee component](https://github.com/luar123/zigbee_esphome) for ESPHome approch their implementations in the future):

* https://github.com/esphome/esphome/pull/10647

**Reply by Hedda** [reply]
By the way, Arduino is now integrated as an ESP-IDF component for ESPHome 2025.10.0 which is a major architectural change(!!!):

* https://esphome.io/changelog/2025.10.0/

"_This release includes a fundamental change in how ESP32 Arduino builds work - **Arduino is now integrated as an ESP-IDF component** rather than a separate framework. **This is the biggest architectural change for ESP32 in ESPHome’s history.**_

_Previously, Arduino came as a precompiled framework with everything baked in - you got whatever upstream decided, whether you needed it or not. This led to bloated builds that **often couldn’t even fit** on devices with limited flash space. You were stuck with the upstream choices and locked out of ESP-IDF features._

_Now Arduino is built as an ESP-IDF component alongside your configuration. ESPHome compiles only what you actually use, eliminating the bloat. You get access to ESP-IDF features while still using Arduino libraries when you need them. This aligns with Espressif’s recommended approach and is how the Arduino ecosystem is evolving._"

...

"_Arduino is still fully supported and works great if you need it for specific Arduino libraries. See our [Arduino to ESP-IDF migration guide](https://esphome.io/components/esp32/#migrating-from-arduino) for help switching - it’s easier than you might think!_"

"_If you maintain external components using Arduino-specific APIs, review [Espressif’s Arduino as ESP-IDF component documentation](https://docs.espressif.com/projects/arduino-esp32/en/latest/esp-idf_component.html) and test your components with this release._"

More info on the other benefits this brings in the ESPHome 2025.10.0 version release notes here as well as in the pull request for it:

* https://esphome.io/changelog/2025.10.0/

  * https://github.com/esphome/esphome/pull/10647

One essential benefit to Zigbee on ESP32 is it add access to latest ESP-IDF 5.4/5.5 features while using Arduino libraries in ESPHome

Note! This framework architecture change with Arduino integrated as an IDF component is a breaking change in ESPHome 2025.10

**Comment by Hedda** [comment]
> _there is also another developer working on Zigbee support for ESHome on ESP32 platform as well so will see if both gets merged, though that other one by luar123 uses a much more low-level approch than what tomaszduda23 is working on for the nRF52 platform._

Update: FYI, [tomaszduda23](https://github.com/tomaszduda23) and [luar123](https://github.com/luar123) are now more aligned and have agreed on how initial Zigbee config with platform/component abstraction should be handled/formatted at a high-level for ESPHome config to be platform-independent, and as such they both each submitted matching pull requests for initial bare-minimum Zigbee support on nRF52 and ESP32 respectively. You can check those out those PRs, see:

* https://github.com/esphome/esphome/pull/11535

and

* https://github.com/esphome/esphome/pull/11553

and

* https://github.com/esphome/esphome-docs/pull/5526

and

* https://github.com/esphome/esphome-docs/pull/5582

_The `zigbee` component allows exposing supported ESPHome components over a Zigbee network to Home Assistant via Zigbee2MQTT** or ZHA. Due to the limitations of the Zigbee protocol, only basic properties are exposed. Additional properties must be configured manually in Home Assistant._

## Example entry for config.yaml in ESPHome:

```
# Example config.yaml
esphome:
  name: zigbee_test

esp32:
  board: esp32-h2-devkitm-1
  framework:
    type: esp-idf

zigbee:

binary_sensor:
  - platform: gpio
    pin:
      number: 9
    name: but1
    report: force
  - platform: gpio
    pin:
      number: 8
    id: but2
    on_click:
      min_length: 5s
      max_length: 20s
      then:
        - zigbee.factory_reset

```

Again, both also still have their previous work open that are more comprehensive but I guess that if the above pull requests are accepted into ESPHome mainline then each of their comprehensive work (linked below) will be rewritten to match the new agreed upon standard:

* https://github.com/esphome/esphome/pull/7340

* https://github.com/luar123/zigbee_esphome

PS: By the way, while it is off-topic you might be interested to know that support for [OpenThread](https://new.esphome.io/components/openthread/) on nrf52 is being worked on for ESPHome, which might be indirectly relevant to this as shows that there are others interested to work on the nrf52 platform in the ESPHome firmware.

**Reply by Hedda** [reply]
@ffenix113 did you check out this pull request for ESPHome? It will probably be merged within a month or sooner as it sounds now which marks the initial support for Zigbee in ESPHome (as well as with OTA makes the nrf52 platform a first-class citizen in ESPHome):

- https://github.com/esphome/esphome/pull/11535

Again, the previous pull request linked below is also still open and more comprehensive but I guess that if the above pull request are accepted into ESPHome mainline then their comprehensive work will be rewritten to match the new agreed upon standard:

- https://github.com/esphome/esphome/pull/7340

PS: Also wanted is for this OTA pull request to be merged as well for completness, see:

- https://github.com/esphome/esphome/pull/11932

and

- https://github.com/esphome/esphome/pull/12344

**Reply by Hedda** [reply]
> Again, the previous pull request linked below is also still open and more comprehensive but I guess that if the above pull request are accepted into ESPHome mainline then their comprehensive work will be rewritten to match the new agreed upon standard:

FYI, [tomaszduda23](https://github.com/tomaszduda23) has now also aubmitted an alternative PR for a refactored Zigbee sensor:

- https://github.com/esphome/esphome/pull/12187

**Comment by Hedda** [comment]
@felipejfc Update: pull requests that added initial Zigbee end device support for the nRF52 platform has now been merged into ESPHome:

- https://github.com/esphome/esphome/pull/11535
  - https://github.com/esphome/esphome/pull/12187
  - https://github.com/esphome/esphome/pull/11535
  - https://github.com/esphome/esphome/pull/11533
  - https://github.com/esphome/esphome/pull/13083
  - https://github.com/esphome/esphome/pull/13114
  - https://github.com/esphome/esphome/pull/12236

That means that the next upcoming ESPHome release (scheduled for release as ESPHome 2026.1.0 in a couple of weeks from now or if that is delayed then in the 2026.2.0 release) will have have initial Zigbee end device support for the nRF52 platform.

- https://next.esphome.io/components/zigbee/
  - https://esphome.io/changelog/2026.1.0/
  - https://esphome.io/changelog/2026.2.0/

Note that the matching pull request that will add initial Zigbee end device support for the ESP32 platform has not yet merged into ESPHome:

- https://github.com/esphome/esphome/pull/11553

Again, also be ware that both [tomaszduda23](https://github.com/tomaszduda23) and [luar123](https://github.com/luar123) also still have their previous work and pulls requests open that offer much more extensive and comprehensive Zigbee device types (which also have much more information and better overview of the project scopes), but I guess that if the two pull requests linked above are accepted and merged into ESPHome mainline then each of their comprehensive work (linked below) will be partially rewritten to align with the new agreed upon standard and formatting. See these for more background information:

- https://github.com/esphome/esphome/pull/7340

and

- https://github.com/luar123/zigbee_esphome

**Comment by Hedda** [comment]
@felipejfc ESPHome 2026.1.0 has now been released with a Zigbee End Device component based on nRF SDK supporting nRF52 platform:

* https://esphome.io/components/zigbee/

"_**The zigbee component allows exposing supported ESPHome components over a Zigbee network to Home Assistant via Zigbee2MQTT or ZHA. Due to the limitations of the Zigbee protocol, only basic properties are exposed. Additional properties must be configured manually in Home Assistant. Each ESPHome entity consumes one Zigbee endpoint. Because of a limitation in Zigbee2MQTT, at least two endpoints are required. The maximum number of supported endpoints is eight.**_"

So far in ESPHome 2026.1 they added initial support however it already features supports for  sensor, binary sensor, and switch functions:

* https://esphome.io/changelog/2026.1.0/

Quote below from the release notes:

## _Zigbee Support for nRF52_ 

_ESPHome now supports Zigbee on nRF52 platforms, bringing a new connectivity option alongside WiFi, Bluetooth, and Thread:_

* _**Binary sensor support** - Expose GPIO states via Zigbee ([#11535](https://github.com/esphome/esphome/pull/11535))_
* _**Sensor support** - Expose ESPHome sensors via Zigbee’s Analog Input cluster with automatic unit mapping ([#12187]_(https://github.com/esphome/esphome/pull/12187))
* _**Switch support** - Control ESPHome switches via Zigbee as binary output ([#13083](https://github.com/esphome/esphome/pull/13083))
wipe_on_boot: once - Wipe network settings only on first boot, preserving connections after OTA updates_
* _**Framework version support** - Configure nRF-SDK version with experimental support for SDK 2.9.2 and 3.2.0 ([#12489](https://github.com/esphome/esphome/pull/12489))_

## Future

Again, Zigbee support in ESPHome is currently available only on nRF52 platforms but there is an open PR to also add ESP32 support as well:

* https://github.com/esphome/esphome/pull/11553

**Reply by ffenix113** [reply]
Super! Glad they did manage to merge and release it! Also, glad that this project (zigbee_home) helped (even a bit) to make it happen.

Options are good. I also want to try and add Matter-over-Thread and allow users to choose which protocol they want to use.

**Reply by Hedda** [reply]
> I also want to try and add Matter-over-Thread and allow users to choose which protocol they want to use.

For reference, ESPHome already support OpenThread and IPv6 on ESP32, as well as matching draft pull requests for the nrf52 platform (which are prerequisites), but as far as I know no one has started working on an Matter application-layer component for ESPHome. See:

* https://github.com/esphome/esphome/pull/16333

and

* https://github.com/esphome/esphome/pull/16336

There is however an open feature request discussion about an Matter application-layer component native in ESPHome here:

* https://github.com/esphome/feature-requests/issues/1430

**Comment by Hedda** [comment]
FYI, ESPHome 2026.5.0 (May 2026) also add Zigbee End Device support to ESP32

https://esphome.io/changelog/2026.5.0/#zigbee-expanded-to-esp32-h2-and-c6

Zigbee component in ESPHome previosuly only supported nrf52

https://esphome.io/components/zigbee/

So now ESPHome has a unified Zigbee component that support both platforms.
